controllers/awsmachinetemplate_controller.go

```diff
@@ -136,201 +134,217 @@ func (r *AWSMachineTemplateReconciler) Reconcile(ctx context.Context, req ctrl.R
 	}
 
 	if awsMachineTemplate.DeletionTimestamp != nil {
-		roleUsed, err := isRoleUsedElsewhere(ctx, r.Client, mainRoleName)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
+		return r.reconcileDelete(ctx, iamService, awsMachineTemplate, logger, clusterName, req.Namespace, role)
+	}
+	return r.reconcileNormal(ctx, iamService, awsMachineTemplate, logger, clusterName, req.Namespace, role)
 
-		if !roleUsed {
-			err = iamService.DeleteRole()
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			if role == iam.ControlPlaneRole {
-				if r.EnableKiamRole {
-					err = iamService.DeleteKiamRole()
-					if err != nil {
-						return ctrl.Result{}, err
-					}
-				}
+}
 
-				if r.EnableRoute53Role {
-					err = iamService.DeleteRoute53Role()
-					if err != nil {
-						return ctrl.Result{}, err
-					}
-				}
-			}
-		}
-		// remove finalizer from AWSCluster
-		{
-			awsCluster, err := key.GetAWSClusterByName(ctx, r.Client, clusterName, awsMachineTemplate.GetNamespace())
-			if err != nil {
-				logger.Error(err, "failed to get awsCluster")
-				return ctrl.Result{}, err
-			}
+func (r *AWSMachineTemplateReconciler) reconcileDelete(ctx context.Context, iamService *iam.IAMService, awsMachineTemplate *capa.AWSMachineTemplate, logger logr.Logger, clusterName, namespace, role string) (ctrl.Result, error) {
+	roleUsed, err := isRoleUsedElsewhere(ctx, r.Client, awsMachineTemplate.Spec.Template.Spec.IAMInstanceProfile)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
 
-			if controllerutil.ContainsFinalizer(awsCluster, key.FinalizerName(iam.ControlPlaneRole)) {
-				patchHelper, err := patch.NewHelper(awsCluster, r.Client)
+	if !roleUsed {
+		err = iamService.DeleteRole()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if role == iam.ControlPlaneRole {
+			if r.EnableKiamRole {
```

Comment:
Didn't we remove `kiam` in CAPA? If that's the case, why not remove all logic related to `kiam` from the operator?

Reply:
I tried it but they are somehow still required (I was not able to create a cluster with them) but we might rename respective parts